pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -422,54 +458,58 @@ func ResolvePipelineRun(
 	state := []*ResolvedPipelineRunTask{}
 	for i := range tasks {
 		pt := tasks[i]
-
 		rprt := ResolvedPipelineRunTask{
 			PipelineTask: &pt,
-			TaskRunName:  GetTaskRunName(pipelineRun.Status.TaskRuns, pt.Name, pipelineRun.Name),
 		}
 
-		// Find the Task that this PipelineTask is using
-		var (
-			t        v1beta1.TaskInterface
-			err      error
-			spec     v1beta1.TaskSpec
-			taskName string
-			kind     v1beta1.TaskKind
-		)
-
-		if pt.TaskRef != nil {
-			if pt.TaskRef.Kind == v1beta1.ClusterTaskKind {
-				t, err = getClusterTask(pt.TaskRef.Name)
+		if rprt.IsCustomTask() {
+			rprt.RunName = GetRunName(pipelineRun.Status.Runs, pt.Name, pipelineRun.Name)
+			run, err := getRun(rprt.RunName)
+			if err != nil && !errors.IsNotFound(err) {
+				return nil, fmt.Errorf("error retrieving Run %s: %w", rprt.RunName, err)
+			}
+			rprt.Run = run
+		} else {
+			rprt.TaskRunName = GetTaskRunName(pipelineRun.Status.TaskRuns, pt.Name, pipelineRun.Name)
+			// Find the Task that this PipelineTask is using
+			var (
+				t        v1beta1.TaskInterface
+				err      error
+				spec     v1beta1.TaskSpec
+				taskName string
+				kind     v1beta1.TaskKind
+			)
+
+			if pt.TaskRef != nil {
+				if pt.TaskRef.Kind == v1beta1.ClusterTaskKind {
+					t, err = getClusterTask(pt.TaskRef.Name)
+				} else {
+					t, err = getTask(pt.TaskRef.Name)
+				}
+				if err != nil {
+					return nil, &TaskNotFoundError{
+						Name: pt.TaskRef.Name,
+						Msg:  err.Error(),
+					}
+				}
+				spec = t.TaskSpec()
+				taskName = t.TaskMetadata().Name
+				kind = pt.TaskRef.Kind
 			} else {
-				t, err = getTask(pt.TaskRef.Name)
+				spec = *pt.TaskSpec.TaskSpec
 			}
+			spec.SetDefaults(contexts.WithUpgradeViaDefaulting(ctx))
+			rtr, err := ResolvePipelineTaskResources(pt, &spec, taskName, kind, providedResources)
 			if err != nil {
-				return nil, &TaskNotFoundError{
-					Name: pt.TaskRef.Name,
-					Msg:  err.Error(),
-				}
+				return nil, fmt.Errorf("couldn't match referenced resources with declared resources: %w", err)
 			}
-			spec = t.TaskSpec()
-			taskName = t.TaskMetadata().Name
-			kind = pt.TaskRef.Kind
-		} else {
-			spec = *pt.TaskSpec.TaskSpec
-		}
-		spec.SetDefaults(contexts.WithUpgradeViaDefaulting(ctx))
-		rtr, err := ResolvePipelineTaskResources(pt, &spec, taskName, kind, providedResources)
-		if err != nil {
-			return nil, fmt.Errorf("couldn't match referenced resources with declared resources: %w", err)
-		}
 
-		rprt.ResolvedTaskResources = rtr
+			rprt.ResolvedTaskResources = rtr
 
-		taskRun, err := getTaskRun(rprt.TaskRunName)
-		if err != nil {
-			if !errors.IsNotFound(err) {
+			taskRun, err := getTaskRun(rprt.TaskRunName)
+			if err != nil && !errors.IsNotFound(err) {
 				return nil, fmt.Errorf("error retrieving TaskRun %s: %w", rprt.TaskRunName, err)
 			}
-		}
-		if taskRun != nil {
 			rprt.TaskRun = taskRun
 		}
 
```

Comment:
There should be a TODO to update `resolveConditionChecks` to understand `PipelineRunRunStatus` as well.

Reply:
Done. All of this is likely going to change with When Expressions too, so it'll be a race! 😄